fast_histogram/_histogram_core.c

```diff
@@ -84,52 +91,113 @@ static PyObject *_histogram1d(PyObject *self, PyObject *args)
 
     /* Build the output array */
     dims[0] = nx;
-    count_array = PyArray_SimpleNew(1, dims, NPY_DOUBLE);
+    count_array = PyArray_SimpleNew(1, dims, NPY_FLOAT64);
     if (count_array == NULL) {
         PyErr_SetString(PyExc_TypeError, "Couldn't build output array");
         Py_DECREF(x_array);
         Py_XDECREF(count_array);
         return NULL;
     }
 
+    /* Releasing GIL, no allocation/referencing allowed
+     (see `ndarrayTypes.h` for macros) */
+    Py_BEGIN_ALLOW_THREADS
     PyArray_FILLWBYTE(count_array, 0);
 
-    /* Get pointers to the data as C-types. */
-
-    x = (double*)PyArray_DATA(x_array);
-    count = (double*)PyArray_DATA(count_array);
+    /* Get pointers to the data as C-Types. */
+    x = (npy_float64*)PyArray_DATA(x_array);
+    count = (npy_float64*)PyArray_DATA(count_array);
 
-    fnx = nx;
+    fnx = (npy_float64)nx;
     normx = 1. / (xmax - xmin);
 
     for(i = 0; i < n; i++) {
+        if (x[i] >= xmin && x[i] < xmax) {
+            ix = (int)((x[i] - xmin) * normx * fnx);
+            count[ix] += 1.;
+        }
+    }
+    
+    /* Clean up. */
+    Py_END_ALLOW_THREADS
+    Py_DECREF(x_array);
 
-      tx = x[i];
```

Comment:
Just to check, I had originally done this for performance reasons to avoid accessing item [i] from ``x`` several times inside the loop - but do you think this doesn't actually provide a performance improvement?

Reply:
No, the compiler will do that for you.